cmake/add_memcheck_target.cmake

```diff
@@ -0,0 +1,25 @@
+# This macro adds a "make memcheck" target that runs Valgrind on all tests.
+# It only works on Linux.
+macro(add_memcheck_target)
+  if (UNIX AND NOT APPLE)
+    find_program(VALGRIND valgrind)
+    if (NOT VALGRIND MATCHES "-NOTFOUND")
+      message(STATUS "Valgrind found. Enabling `make memcheck`")
+      set(VALGRIND_FOUND 1)
+      set(CTEST_MEMORYCHECK_COMMAND ${VALGRIND})
+      # Add "--gen-suppressions=all" to MEMORYCHECK_COMMAND_OPTIONS to generate
+      # suppressions for Valgrind's false positives. The suppressions show up
+      # right in the MemoryChecker.*.log files.
+      set(CTEST_MEMORYCHECK_COMMAND_OPTIONS "--leak-check=full --show-leak-kinds=all --errors-for-leak-kinds=definite,possible --track-origins=yes --error-exitcode=1 --trace-children=yes --suppressions=${PROJECT_SOURCE_DIR}/tools/valgrind/scasm.supp" CACHE STRING "Options passed to Valgrind." FORCE)
+
+      # make memcheck target
+      add_custom_target(memcheck ctest -T memcheck -j USES_TERMINAL)
+    else()
+      set(VALGRIND_FOUND 0)
+    endif()
+  else()
+    # Valgrind doesn't work on Macs.
+    set(VALGRIND_FOUND 0)
+  endif()
```

Comment:
This adds a custom target, but does it run any tests? Are these memcheck tests meant to be running somewhere in GitHub actions or PNNL actions?

I couldn't find anything in the logs from looking at it :(

Reply:
This adds a custom target and teaches `ctest` how to run with the `-T memcheck` flag. Because Valgrind slows everything way down, we don't actually use these tests in CI, and I wouldn't suggest we do unless we do some pruning. So for now, this is a "run on your machine" target.

Reply:
We couldn't have say one GitHub action exclusively run Valgrind tests for one build type?

My experience with checks that we hope people remember to run (as well as run on a non-MacOS) is that they often forget.

Reply:
We could do that, but not with our current batch of tests--they just take too long. I don't think it would be difficult to prune the validation tests so they'd run faster, and once we do that, we can add a memcheck test.